File: proposed-projects/round-trip-experiment/reverse/20260909T015142Z_f8ca89ff/20260910T053604Z_b86749e4/ported_implementation/src/rules-builder/parse-name.ts
import { GrammarParseError } from '../utils/errors/grammar-parse-error';

export const PARSE_NAME_ERROR = 'Failed to find a valid name';

export const GET_INVALID_CHAR_ERROR = (char: string): string =>
  `Invalid character "${char}" when parsing name, only lowercase letters and hyphens are allowed.`;

const VALID_CHAR = /[a-zA-Z-]/;
const INVALID_NEXT_CHAR = /[_0-9]/;

const isValidChar = (char: string): boolean => VALID_CHAR.test(char);
const isInvalidNextChar = (char: string): boolean => INVALID_NEXT_CHAR.test(char);

export const parseName = (grammar: string, pos: number): string => {
  let name = '';
  while (pos < grammar.length && isValidChar(grammar[pos])) {
    name += grammar[pos];
    pos += 1;
  }
  if (!name) {
    throw new GrammarParseError(grammar, pos, PARSE_NAME_ERROR);
  }
  if (pos < grammar.length && isInvalidNextChar(grammar[pos])) {
    throw new GrammarParseError(grammar, pos, GET_INVALID_CHAR_ERROR(grammar[pos]));
  }
  return name;
};

File: packages/gbnf-experiment/data/20260909T134433Z_3f62598c/ported_implementation/gbnf/rules_builder/parse_name.py
import re

from ..utils.errors.grammar_parse_error import GrammarParseError

PARSE_NAME_ERROR = "Failed to find a valid name"


def GET_INVALID_CHAR_ERROR(char: str) -> str:
    return (
        f'Invalid character "{char}" when parsing name, '
        "only lowercase letters and hyphens are allowed."
    )


_VALID_CHAR = re.compile(r"[a-zA-Z-]")
_INVALID_NEXT_CHAR = re.compile(r"[_0-9]")


def parse_name(grammar: str, pos: int) -> str:
    name = ""
    while pos < len(grammar) and _VALID_CHAR.search(grammar[pos]):
        name += grammar[pos]
        pos += 1
    if not name:
        raise GrammarParseError(grammar, pos, PARSE_NAME_ERROR)
    if pos < len(grammar) and _INVALID_NEXT_CHAR.search(grammar[pos]):
        raise GrammarParseError(grammar, pos, GET_INVALID_CHAR_ERROR(grammar[pos]))
    return name
